subsys/fs/nvs/nvs.c

```diff
@@ -6,104 +6,492 @@
  */
 
 #include <flash.h>
-#include <crc16.h>
 #include <string.h>
 #include <errno.h>
 #include <inttypes.h>
 #include <nvs/nvs.h>
 #include "nvs_priv.h"
 
+/* basic calculation routines */
 
-static inline u16_t _nvs_len_in_flash(struct nvs_fs *fs, u16_t len)
+static size_t _nvs_flash_align(struct nvs_fs *fs, size_t len)
 {
 	if (fs->write_block_size <= 1) {
 		return len;
 	}
 	return (len + (fs->write_block_size - 1)) & ~(fs->write_block_size - 1);
 }
 
-u16_t _nvs_entry_len_in_flash(struct nvs_fs *fs, u16_t len)
+size_t _nvs_flash_size(struct nvs_fs *fs, size_t len)
 {
-	return _nvs_len_in_flash(fs, len) +
-	       _nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr)) +
-	       _nvs_len_in_flash(fs, sizeof(struct _nvs_data_slt));
+	if (len < 0x80) {
+		return _nvs_flash_align(fs, len + sizeof(u16_t)
+					+ 2*sizeof(u8_t));
+	}
+	return _nvs_flash_align(fs, len + sizeof(u16_t)
+					+ 2*sizeof(u16_t));
 }
 
+void _nvs_addr_advance(struct nvs_fs *fs, off_t *addr, size_t step)
```

Comment:
I think all those private function that are not part of the API should just be marked as static in `nvs.c,` so that the compiler can do more aggressive optimizations. I just report that here, but that applies to many functions of this file.